Migrate RCAttributionTypeFactory

Juanpe opened the pull request:
Resolves #574 

The goal of this PR is to migrate the `RCAttributionTypeFactory` object to `Swift`. I added some `// TODO` to make it internal when the migration is done.

About the ROT13 algorithm, I did copy/paste from this [implementation](https://www.hackingwithswift.com/example-code/strings/how-to-calculate-the-rot13-of-a-string). Let me know if it's valid for you or if you need to use a specific version of this algorithm.

aboedo commented:
Hi @Juanpe!! Thanks so much for opening! 🙌 

I think this class is trickier than usual since the objective-c version did some trickery to get things to work that won't translate well to Swift. I'll leave detailed notes in the code

aboedo commented on Purchases/Attribution/RCAttributionTypeFactory.m:
```diff
@@ -1,78 +0,0 @@
-    // so that Apple can't find them during the review, but we can still access them on runtime.
-    NSString *className = [self rot13:self.mangledTrackingClassName];
-
-    return (Class<FakeTrackingManager> _Nullable)NSClassFromString(className);
```
this line from the old code may seem straightforward, but what it's actually doing is the tricky part - it's relying on the fact that `FakeTrackingManager` exposes the same methods we're looking for in `ATTrackingManager` to call the same methods. 
Since in objc this really means passing a message with the same name and values, it'll work as long as the interface of `FakeTrackingManager` matches whatever we're trying to use from `ATTrackingManager`

taquitos replied:
I was intrigued by this, and having spent a good deal of time mucking around with the objc runtime, I wanted to give it a go with Swift. Turns out, it's not too bad! Just not as clever as what currently exists:

Since we're working with Apple's ObjC classes, that means we still have NSObject subclasses that we're dealing with. So, we can do some dynamic dispatch 🕺

```swift
guard let klass = NSClassFromString("ASIdentifierManager"), let myClass = klass as? NSObjectProtocol else {
    // ASIdentifierManager hasn't been loaded in the ObjC runtime yet
    print("sad panda")
    return
}

let sharedManager = myClass.perform(Selector("sharedManager")).takeRetainedValue()
let uuid = sharedManager.perform(Selector("advertisingIdentifier")).takeRetainedValue()

```

taquitos replied:
![Screen Shot 2021-07-09 at 6 29 17 PM](https://user-images.githubusercontent.com/1304821/125147917-93650000-e0e3-11eb-8987-1183ff8051ed.png)

aboedo replied:
great idea! @Juanpe I think this approach would work. We'd still need to retain the ROT13 mangling, of course. 

@taquitos could you clarify the bit about referencing the class in code? Currently this works even if the class isn't referenced, as long as the framework has been linked in the build phases. 
I think if you don't link it, then this only works if you've referenced the class, a behavior that started in iOS 14.something (can't recall the exact release). 

Could this be what you're seeing?

taquitos replied:
I'm 99% sure that's exactly it- it must be linked. Last I knew, linking happens when a class is referenced at least once in the codebase. That is, the compiler hasn't optimized it out. This is a problem for ObjC where a framework includes categories, but the app using the framework doesn't use the category. The compiler would strip it out unless you used the `-objc` linker flag to retain those things. In the past it was not enough to have the library/framework linked in the xcode project, it needed to be referenced somewhere in the app's code.

taquitos replied:
The good news, I think this is relatively easy to check for- build an app and run the code.

aboedo commented on PurchasesCoreSwift/Attribution/AttributionTypeFactory.swift:
```diff
@@ -0,0 +1,72 @@
+        // functions in ATTrackingTransparency. We apply rot13 on these functions and classes names
+        // so that Apple can't find them during the review, but we can still access them on runtime.
+        let className = mangledTrackingClassName.rot13()
+        return NSClassFromString(className) as? FakeTrackingManager.Type
```
This is the bit that won't work in Swift - even though the class is correctly obtained in `NSClassFromString`, it won't be castable to `FakeTrackingManager`, since the class doesn't explicitly conform to the protocol. 
Swift doesn't have reflection, so we can't call the method by name either. 

I'm honestly not sure what the best fix for this is, would love suggestions

aboedo replied:
Here are some steps to reproduce: 

Here are some testing steps to help figure this out: 

In an app that integrates the RevenueCat SDK (you can use the one in `IntegrationTests/PurchaseTester`, you'll have to run `pod install` and update the signing team to your own).
- in Build Phases, in Link Binary with Libraries, add `AdServices` and `AppTrackingTransparency`. 
- in the app, on app launch, add the following code. This will trigger a tracking request, make sure you select Allow. If the tracking request doesn't show up, make sure your device doesn't have tracking disabled or enable it in `Settings -> Privacy -> Tracking -> Allow Apps to Request to Track`
```swift
ATTrackingManager.requestTrackingAuthorization { status in
    Purchases.automaticAppleSearchAdsAttributionCollection = true
    Purchases.configure(withAPIKey: <your_public_sdk_key_here>)
    Purchases.logLevel = .debug
}
```

This will trigger: a modal to track (again, make sure you accept), then a call to these methods to figure out whether the frameworks are linked in the app, instantiate their classes and call their corresponding methods. 
Make sure you set a breakpoint here to see if the class can be obtained with `NSClassFromString(className)`, and then casted with `as? FakeTrackingManager.Type`.

Juanpe replied:
Ohh, yeah, I can see the problem 🤔 I've checked the objc runtime API but I can't find a valid solution for that...Firstly, I thought that maybe we could replace the implementation of the fake method with the original one by swizzling but I couldn't find the way.

I've checked also the idea of using the `unsafeBitCast` method to create an empty class with the same size as `ATTrackingManager` and replace the implementation of the methods instead of using a `FakeProtocol` but I was facing an error as well :/

It seems that the solution provided by @taquitos could be useful, right?

BTW, I think there is another approach but it's harder for the developers. The idea is inversing the control, I mean, we only need to know if the tracking is authorized or not, so we could use a closure or a delegate method to ask to the developer code the current status. The code would look like this:

```swift
ATTrackingManager.requestTrackingAuthorization { status in
    Purchases.trackingAuthorizationStatusProvider = <Closure or delegate here>
    Purchases.automaticAppleSearchAdsAttributionCollection = true
    Purchases.configure(withAPIKey: <your_public_sdk_key_here>)
    Purchases.logLevel = .debug
}
```

The problem with this approach is the migration would change the logic and the API and I read in the Swift migration plan that you only want to do a translation from ObjC to Swift.

Juanpe replied:
Hi again, I think I found a possible way to do that.

As [Apple's doc](https://developer.apple.com/documentation/swift/anyclass) says if you use`AnyClass`: 
> _"all known @objc class methods and properties are available as implicitly unwrapped optional methods and properties"_

So, if we use a `class` instead of a `protocol` to define our `FaceTrackingManager` and we add all the methods that we need. Imagine this definition:

```swift
class FakeTrackingManager {
    @objc class func trackingAuthorizationStatus() -> Int {
        -1
    }
    
    @objc class func requestTrackingAuthorization(completionHandler completion: @escaping (Int) -> Void) {
        // do nothing
    }
}
```

We could create a "`proxy`", something like `TrackingProxy` that perform the methods that we need. For example:
```swift
class TrackingProxy {
    static func trackingAuthorizationStatus(_ trackingClass: AnyClass) -> Int {
        trackingClass.trackingAuthorizationStatus()
    }
    
    static func requestTrackingAuthorization(_ trackingClass: AnyClass, completion: @escaping (Int) -> Void) {
        trackingClass.requestTrackingAuthorization(completionHandler: completion)
    }
}
```

Then, we only need to initialize the class type, `ATTrackingManager` from the mangled string, and pass it as param in the proxy methods:

```swift
let classType: AnyClass = NSClassFromString(mangledTrackingClassName.rot13()) ?? FakeTrackingManager.self
        
TrackingProxy.requestTrackingAuthorization(classType, completion: { status in
                                                        print(status)
                                                   })
```

**Sample app**
<img width="200px" src="https://user-images.githubusercontent.com/1409041/125303511-2775d800-e32d-11eb-9e0e-d8f904ab52f9.png"/>

If we tap on **"Ask App Not to Track"** and execute this code:
```swift
let classType: AnyClass = NSClassFromString(mangledTrackingClassName.rot13()) ?? FakeTrackingManager.self
print(TrackingProxy.trackingAuthorizationStatus(classType)) // 2
```
 the status is: `denied`

WDYT? Do you think this could be useful for us?

taquitos replied:
This looks even smarter than what I suggested!

Juanpe replied:
Great! if it looks good for you, I'll make a proposal using this approach, ok?

taquitos replied:
Yeah, it looks great to me 🤩

aboedo reviewed (commented):
Some more context on the intent of the stuff in this class: 

We've had apps get rejected by the app review team, because the app review team said they were using tracking stuff... but they weren't at all, and didn't even have the tracking frameworks linked in the app. It seems like App Review uses some grep to find the class names, so we ended up mangling the class names to get around that. 

This is what this class does, but figuring out an equivalent in Swift is tricky for sure. 

I'd love to hear your thoughts on this! 

And let me know if the repro steps aren't working or you need a hand.

Juanpe commented:
Hi!

I did the changes taking into account the approach that I told you about. Now, the factory makes proxies instead of the original classes. Then, this proxy uses the `AnyClass` features to call dynamically the method that we need.

Moreover, I've added some comments to clarify why we need some empty methods, and I've added the prefix `RC` to the proxies.

Let me know if I should do more changes 🙂

taquitos commented on PurchasesCoreSwift/Attribution/AttributionTypeFactory.swift:
```diff
@@ -0,0 +1,123 @@
+
+    // We need this method to be available as an optional implicitly unwrapped method for `AnyClass`.
+    @objc func requestAttributionDetails(_ completionHandler: AttributionDetailsBlock) {
+        // do nothing
```
Would it make sense to add some logging here, like a warning that this method was called and it shouldn't have been?

Juanpe replied:
Well, this method would execute if the original class couldn't be instantiated. If so, the factory does an early return and log a message indicating that the framework hasn't been imported yet. 

But, TBH, I think if we add a warning message, it's not too much, and It could give us more context for weird scenarios 👍🏼

taquitos commented on PurchasesCoreSwift/Attribution/AttributionTypeFactory.swift:
```diff
@@ -0,0 +1,123 @@
+    case authorized
+}
+
+class FakeAdClient: NSObject {
```
Add a comment about why this class exists 📔  so future contributors aren't like "Lolwut?"

Juanpe replied:
Makes sense 🙂

taquitos commented on PurchasesCoreSwift/Attribution/AttributionTypeFactory.swift:
```diff
@@ -0,0 +1,123 @@
+//
+//  AttributionTypeFactory.swift
```
Let's split the classes out of this file into their own 
📖 -> 📃📃📃

Juanpe replied:
Yep, agree with you

taquitos commented on PurchasesTests/Mocks/MockAttributionTypeFactory.swift:
```diff
@@ -24,33 +17,33 @@ class MockAdClient: NSObject, FakeAdClient {
-
-    static func trackingAuthorizationStatus() -> Int {
-        return Int(mockAuthorizationStatus.rawValue)
+    
```
extra spaces

taquitos reviewed (changes requested):
Tiny changes, but basically ready to 🚢

Juanpe commented:
Hi! 
Thanks for the feedback, I did the changes that you suggested to me 👍🏼 

I've added some comments explaining why we need the fake classes, let me know if it's clear or if we should add more info to give more context to future contributors :)

taquitos reviewed (commented):
Looks good to me, I'll defer to @aboedo's review 😄 
Thank you @Juanpe!

taquitos commented:
I haven't had an opportunity to test this, and my brain is fried right now, I'll check back in tomorrow and test it out.

taquitos commented:
Same story today, I thought https://github.com/RevenueCat/purchases-ios/pull/658 would take a lot less time. I'm hoping to track down my last remaining problems and then get back to checking this pr!

Juanpe commented:
> Same story today, I thought #658 would take a lot less time. I'm hoping to track down my last remaining problems and then get back to checking this pr!

no problem 😉

taquitos commented on PurchasesCoreSwift/Attribution/AdClientProxy.swift:
```diff
@@ -0,0 +1,44 @@
+
+// TODO(Post-migration): switch this back to internal the class and all these protocols and properties.
+
+public typealias AttributionDetailsBlock = ([String: Any]?, Error?) -> Void
```
I think this should be `([String : NSObject]?, Error?) -> Void`
(Any -> NSObject)
I'm checking the header from the iAd API:
![Screen Shot 2021-07-20 at 2 35 00 PM](https://user-images.githubusercontent.com/1304821/126398483-70eea2b5-ff40-492f-a7e2-2b0929ca25cd.png)

Juanpe replied:
makes more sense 👍🏼 

I've done the [change](https://github.com/RevenueCat/purchases-ios/pull/635/commits/c65f70fecb8092644e835774267fc7e45cf8f8f4) 🙂

taquitos commented:
Sorry for the delay, in testing this, I'm seeing a crash whenever I attempt to use `client.requestAttributionDetails`
on the `AdClientProxy`. It looks like some residual objc/swift bridging issues. It could potentially be a Swift bug since if I replace the `client` with `ADClient.shared()` and pass the same block, it works just fine. There are a couple differences, one is the method signature we should use is:
`open func requestAttributionDetails(_ completionHandler: @escaping ([String : NSObject]?, Error?) -> Void) {`
The (note the addition of `escaping`)

That doesn't fix the issue, though.
I also noticed I can't call any other functions without the same crash. Like if I add:
`open func lookupAdConversionDetails(_ completionHandler: @escaping (Date?, Date?) -> Void) {`
and try to call that, same crash.

taquitos commented:
@Juanpe I figured it out! Check out my commit ^

Juanpe commented:
Sorry, I took a few days off  🙂
Thanks for the commit. TBH, it's so weird 😅 

Just to know, the unretained method creates a weak reference, doesn't it?

taquitos commented:
@Juanpe yeah, I tried it with both unretained and retained method. I think because it's a singleton, the `iAd` framework actually does maintain +1 retain count on it.

**Re: super weird:** Yeah, it's SUPER WEIRD. Like, it should work, but it just crashes. 

Also re: time off, me too 😄

taquitos commented on Purchases.xcodeproj/project.pbxproj:
```diff
@@ -724,22 +728,14 @@
-				37E3564466002C9162AC7C5E /* Package.swift */,
-				B3AA35A2268683540007771B /* EntitlementInfo.swift */,
-			);
-			name = "Recovered References";
```
I hate merging `.xcodeproj`

Juanpe replied:
same here 

![](https://media.giphy.com/media/7T33BLlB7NQrjozoRB/giphy-downsized.gif)

taquitos reviewed (commented):
@RevenueCat/sdk ready to go!

beylmk commented on PurchasesCoreSwift/Attribution/TrackingManagerProxy.swift:
```diff
@@ -0,0 +1,66 @@
+
+}
+
+// We need this class to avoid Kid apps being rejected for getting idfa. It seems like App
```
wondering whether we should be clear in these comments that we won't actually try to improperly access idfa 😅

Juanpe replied:
It's a good question 🤔 I think this comment aims to clarify why this class exists for future contributors. So, I'm not sure if we should include here that we won't actually try to improperly access idfa. I think this info should be included in the doc of the library, shouldn't it? 
Although I don't have a strong opinion on that, so if we decide to include it, that's fine by me 🙂

WDTY?

taquitos replied:
Definitely for future contributors, and maybe folks who are just curious if we're doing anything with idfa? People who might be auditing, etc. I'll add a little message because I think it's nice to point out.

Juanpe replied:
Thanks @taquitos :)

aboedo commented on Purchases/Attribution/RCAttributionFetcher.h:
```diff
@@ -8,13 +8,14 @@
+// TODO: remove it when the attribution fetcher will be migrated
+typedef void (^RCAttributionDetailsBlock)(NSDictionary<NSString *, NSObject *> *_Nullable, NSError *_Nullable);
+
+@class RCDeviceCache, RCIdentityManager, RCBackend, RCAttributionTypeFactory;
+@class RCSystemInfo;
```
suuper minor, but these could all live in a single line, right?

Juanpe replied:
sure :)

aboedo reviewed (approved):
looks fantastic! I also tested it locally and it's working correctly. 

Thanks for sticking with us @Juanpe!! Amazing work! 🎉 

Also shoutout to @taquitos for helping come up with workarounds!

taquitos reviewed (approved):
🥅    ⚽      🐐
